src/py/flwr/cli/log.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Flower command line interface `log` command."""
+
+import sys
+import time
+from logging import DEBUG, ERROR, INFO
+from pathlib import Path
+from typing import Dict, Optional
+
+import grpc
+import typer
+from typing_extensions import Annotated
+
+from flwr.cli.config_utils import load_and_validate
+from flwr.common.grpc import GRPC_MAX_MESSAGE_LENGTH, create_channel
+from flwr.common.logger import log as logger
+from flwr.proto.exec_pb2 import StreamLogsRequest  # pylint: disable=E0611
+from flwr.proto.exec_pb2_grpc import ExecStub
+
+CONN_REFRESH_PERIOD = 60  # Connection refresh period for log streaming (seconds)
+
+
+def stream_logs(run_id: int, channel: grpc.Channel, duration: int) -> None:
+    """Stream logs from the beginning of a run with connection refresh."""
+    start_time = time.time()
+    stub = ExecStub(channel)
+    req = StreamLogsRequest(run_id=run_id)
+
+    for res in stub.StreamLogs(req):
+        print(res.log_output)
+        if time.time() - start_time > duration:
+            break
+
+
+def print_logs(run_id: int, channel: grpc.Channel, timeout: int) -> None:
+    """Print logs from the beginning of a run."""
+    stub = ExecStub(channel)
+    req = StreamLogsRequest(run_id=run_id)
+
+    while True:
+        try:
+            # Enforce timeout for graceful exit
+            for res in stub.StreamLogs(req, timeout=timeout):
+                print(res.log_output)
+        except grpc.RpcError as e:
+            # pylint: disable=E1101
+            if e.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
+                break
+            if e.code() == grpc.StatusCode.NOT_FOUND:
+                logger(ERROR, "Invalid run_id `%s`, exiting", run_id)
+                break
+            if e.code() == grpc.StatusCode.CANCELLED:
+                break
+
+    channel.close()
```

Comment:
My bad! I missed the while loop. Maybe we can wrap it in another try, finally?
e.g.:
```python
try:
    while True:
        try:
            fun()
        except Exception:
            break
finally:
    channel.close()
```

Reply:
@Robert-Steiner The reason `channel.close()` is outside of the `try-except` loop is to close the channel when the connection hits the `timeout` on this line:
```
for res in stub.StreamLogs(req, timeout=timeout):
...
```

So, when we only print the logs with `flwr log -F -- <RUN_ID> <APP_DIR> <FEDERATION_NAME>`, the `print_logs()` function will display the logs and wait until the `timeout` is hit and then it goes to `channel.close()` and returns to shell. We should not need to hit CTRL-C here.

Reply:
This is slightly clunky, but the reason is because we are using the streaming RPC response for printing.

Reply:
Another possibly cleaner option is to break from the `yield StreamLogsResponse(log_outputs=logs[i])` if it receives a `print` request. Something like:

```python
file: exec_servicer.py
# Other code
for i in range(last_sent_index, len(logs)):
    yield StreamLogsResponse(log_output=logs[i])
last_sent_index = len(logs)
if request.print:
    context.cancel()
```

Wdyt @Robert-Steiner @tanertopal?

Reply:
I'm not so familiar with the code but i like your idea of having the gRPC server notify/signal the client when there are no more messages remaining, rather than relying on a client-side timeout.

I still would suggest wrapping the while loop in a try/finally block to guarantee that the channel is closed in the event of an exception, ensuring a clean exit regardless of the outcome. The user may want to terminate the process if the server sends an excessive amount of logs.

Reply:
Good idea, will adopt this.